backport/datasync/datasync.py

```diff
@@ -0,0 +1,335 @@
+        a = client.get_object(
+            Bucket=S3_BUCKET,
+            Key=_dataset_success_s3_key(dataset_id),
+        )
+        d = json.loads(a["Body"].read().decode())
+        d["dataEditedAt"] = dt.datetime.utcfromtimestamp(d["dataEditedAt"] / 1000)
+        d["metadataEditedAt"] = dt.datetime.utcfromtimestamp(d["metadataEditedAt"] / 1000)
+        d.pop("name", None)
+        return cls(**d)
+
+    def save_to_s3(self, client):
+        client.put_object(
+            Bucket=S3_BUCKET,
+            Body=pd.Series(self.to_dict()).to_json(),
+            Key=_dataset_success_s3_key(self.id),
+            ContentType="application/json",
+            ACL="public-read",
+        )
+
+    def matches(self, ds: "DatasetSync") -> bool:
+        # compare timestamps of the latest update (we don't have sourceChecksum for all datasets)
+        # NOTE: we rebake both data and metadata even if only one has changed. This is a bit inefficient, but should
+        # not matter much since dataset updates are rare.
+        if self.dataEditedAt == ds.dataEditedAt and self.metadataEditedAt == ds.metadataEditedAt:
+            # checksums should match if dataEditedAt match
+            assert self.sourceChecksum is None or ds.sourceChecksum is None or self.sourceChecksum == ds.sourceChecksum
+            return True
+        else:
+            return False
+
+
+def _variable_data_s3_key(variable_id: int) -> str:
+    return f"{S3_PREFIX}/data/{variable_id}.json"
+
+
+def _variable_metadata_s3_key(variable_id: int) -> str:
+    return f"{S3_PREFIX}/metadata/{variable_id}.json"
+
+
+def _dataset_success_s3_key(dataset_id: int) -> str:
+    return f"{S3_PREFIX}/success/_success_dataset_{dataset_id}"
+
+
+def _variable_dataPath(variable_id: int) -> str:
+    return os.path.join(S3_ENDPOINT, _variable_data_s3_key(variable_id))
+
+
+def _variable_metadataPath(variable_id: int) -> str:
+    return os.path.join(S3_ENDPOINT, _variable_metadata_s3_key(variable_id))
+
+
+def _load_variable_ids(engine: Engine, dataset_id: int) -> List[int]:
+    q = """
+    select id from variables where datasetId = %(dataset_id)s
+    """
+    return pd.read_sql(q, engine, params={"dataset_id": dataset_id})["id"]
+
+
+def _load_datasets(engine: Engine, dataset_ids: tuple[int], dt_start: Optional[dt.datetime]) -> List[DatasetSync]:
+    if dataset_ids:
+        where = "id in %(dataset_ids)s"
+    elif dt_start:
+        where = "dataEditedAt > %(dt_start)s or metadataEditedAt > %(dt_start)s"
+    else:
+        # load datasets with at least one chart
+        where = """
+        (
+            -- must be used in at least one chart
+            id in (
+                select distinct v.datasetId from chart_dimensions as cd
+                join variables as v on cd.variableId = v.id
+            )
+        )
+        -- we would like to exclude archived datasets, but we cannot do it
+        -- because some charts still use them (e.g. dataset 88)
```

Comment:
I don't understand why archived datasets should be included (and I don't see any charts using [dataset 88](https://owid.cloud/admin/datasets/88)).

Reply:
Uh, maybe it was a different dataset. Anyway, I'll remove this part of code. We need to be syncing all datasets, even the archived ones (because even archived datasets can have charts).

Reply:
Fair enough, although I don't understand why a dataset would be archived if it still contains active charts.

Reply:
> Fair enough, although I don't understand why a dataset would be archived if it still contains active charts.

Neither do I. I'll see if I can find it and create an issue for it.